src/WrappedMToken.sol

```diff
@@ -502,38 +541,29 @@ contract WrappedMToken is IWrappedMToken, Migratable, ERC20Extended {
 
     /**
      * @dev   Increments total earning supply by `amount_` tokens.
-     * @param amount_       The present amount of tokens to increment total earning supply by.
-     * @param currentIndex_ The current index used to compute the principal amount.
+     * @param amount_    The present amount of tokens to increment total earning supply by.
+     * @param principal_ The principal amount of tokens to increment total earning principal by.
      */
-    function _addTotalEarningSupply(uint240 amount_, uint128 currentIndex_) internal {
+    function _addTotalEarningSupply(uint240 amount_, uint112 principal_) internal {
         unchecked {
             // Increment the total earning supply and principal proportionally.
             totalEarningSupply += amount_;
-            principalOfTotalEarningSupply += IndexingMath.getPrincipalAmountRoundedUp(amount_, currentIndex_);
+            totalEarningPrincipal = UIntMath.safe112(uint256(totalEarningPrincipal) + principal_);
         }
     }
 
     /**
      * @dev   Decrements total earning supply by `amount_` tokens.
-     * @param amount_       The present amount of tokens to decrement total earning supply by.
-     * @param currentIndex_ The current index used to compute the principal amount.
+     * @param amount_    The present amount of tokens to decrement total earning supply by.
+     * @param principal_ The principal amount of tokens to decrement total earning principal by.
      */
-    function _subtractTotalEarningSupply(uint240 amount_, uint128 currentIndex_) internal {
-        if (amount_ >= totalEarningSupply) {
-            delete totalEarningSupply;
-            delete principalOfTotalEarningSupply;
-
-            return;
-        }
-
-        uint112 principal_ = IndexingMath.getPrincipalAmountRoundedDown(amount_, currentIndex_);
+    function _subtractTotalEarningSupply(uint240 amount_, uint112 principal_) internal {
+        uint240 totalEarningSupply_ = totalEarningSupply;
+        uint112 totalEarningPrincipal_ = totalEarningPrincipal;
 
         unchecked {
-            principalOfTotalEarningSupply -= (
-                principal_ > principalOfTotalEarningSupply ? principalOfTotalEarningSupply : principal_
-            );
-
-            totalEarningSupply -= amount_;
+            totalEarningSupply = totalEarningSupply_ - UIntMath.min240(amount_, totalEarningSupply_);
```

Comment:
a lot of checks: this one should have already covered principal https://github.com/m0-foundation/wrapped-m-token/pull/102/files#diff-82f33536a281f203780891834e2f2453a322bed686e89ea9b7171a2d74fc10b9R442-R448

Reply:
That link does not seemt o work do I do not know what you are referring to. In any case, the `UIntMath.min240` and `UIntMath.min112` here just prevent underflows without the `_subtractTotalEarningSupply` function needing to be aware of how its callers operate or check. It is a safety measurew,e specially if this function will be used again later down the road. It is minimal gas for that safety.